soroban-sdk/src/muxed_address.rs

```diff
@@ -0,0 +1,316 @@
+                        .unwrap_infallible(),
+                )
+                .unwrap(),
+            ),
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn unchecked_new_from_address(env: Env, obj: AddressObject) -> Self {
+        Self {
+            env,
+            obj: AddressObjectWrapper::Address(obj),
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn unchecked_new(env: Env, obj: MuxedAddressObject) -> Self {
+        Self {
+            env,
+            obj: AddressObjectWrapper::MuxedAddress(obj),
+        }
+    }
+
+    #[inline(always)]
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    pub fn as_val(&self) -> &Val {
+        match &self.obj {
+            AddressObjectWrapper::Address(o) => o.as_val(),
+            AddressObjectWrapper::MuxedAddress(o) => o.as_val(),
+        }
+    }
+
+    pub fn to_val(&self) -> Val {
+        match self.obj {
+            AddressObjectWrapper::Address(o) => o.to_val(),
+            AddressObjectWrapper::MuxedAddress(o) => o.to_val(),
+        }
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScVal> for MuxedAddress {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: &ScVal) -> Result<Self, Self::Error> {
+        let v = Val::try_from_val(env, val)?;
+        match val {
+            ScVal::Address(sc_address) => match sc_address {
+                ScAddress::Account(_) | ScAddress::Contract(_) => {
+                    Ok(AddressObject::try_from_val(env, &v)?
+                        .try_into_val(env)
+                        .unwrap_infallible())
+                }
+                ScAddress::MuxedAccount(_) => Ok(MuxedAddressObject::try_from_val(env, &v)?
+                    .try_into_val(env)
+                    .unwrap_infallible()),
+                ScAddress::ClaimableBalance(_) | ScAddress::LiquidityPool(_) => {
+                    panic!("unsupported ScAddress type")
+                }
+            },
+            _ => panic!("incorrect scval type"),
+        }
+    }
+}
+
+#[cfg(not(target_family = "wasm"))]
+impl TryFromVal<Env, ScAddress> for MuxedAddress {
+    type Error = ConversionError;
+    fn try_from_val(env: &Env, val: &ScAddress) -> Result<Self, Self::Error> {
+        ScVal::Address(val.clone()).try_into_val(env)
+    }
+}
+
+#[cfg(any(test, feature = "testutils"))]
+#[cfg_attr(feature = "docs", doc(cfg(feature = "testutils")))]
+impl crate::testutils::MuxedAddress for MuxedAddress {
+    fn generate(env: &Env, id: u64) -> crate::MuxedAddress {
+        let sc_val = ScVal::Address(crate::env::internal::xdr::ScAddress::MuxedAccount(
+            crate::env::internal::xdr::MuxedEd25519Account {
+                ed25519: crate::env::internal::xdr::Uint256(
+                    env.with_generator(|mut g| g.address()),
+                ),
+                id,
+            },
+        ));
+        sc_val.try_into_val(env).unwrap()
+    }
```

Comment:
One last feedback on this function: I think we should drop the `id` parameter.

It feels odd to me that you can generate and provide part of the address as a fixed value. 

Conceptually most devs who are working with muxed addresses have a fixed base address, and variable ID, so having the reverse in this function, a random base and fixed ID, is conceptually a bit mind bending.

If folks really need to set the ID, they can call generate, then use `new` to change the id to a fixed value.